### PR TITLE
Improve tokenizer handling and optimize training process

### PR DESCRIPTION
**Pull request by MaizeHH**
This PR: 
- Handles arrow data with/without spaces. 
- Applies padding until maximum text length is reached. 
- Applies a WSD learning rate scheduler to training. 
- Checkpoints more frequently (per N steps). 
- Better logging under training, showing dynamic learning rate and loss. 
- Applies inference params for faster inference. 
- Runs prediction until corresponding cipher length is hit.

### CODE REVIEW

File: src/engine/solver.py [hunk 1]
```diff
@@ -78,39 +79,66 @@ def load_checkpoint(self, checkpoint_path: Path) -> "CipherSolver":
...
+
+        sep_id = self.tokenizer.sep_token_id
+        if sep_id in input_ids[0]:
+            sep_idx = (input_ids[0] == sep_id).nonzero(as_tuple=True)[0][0]
+            input_ids = input_ids[:, :sep_idx + 1]
+        else:
+            sep_tensor = torch.tensor([[sep_id]], device=self.device)
+            input_ids = torch.cat([input_ids, sep_tensor], dim=1)
+
+        cipher_len = input_ids.size(1) - 2
```
Comment (Jatewo) [inline src/engine/solver.py]: This line seems off to me. Can you explain why the `-2`?

File: src/data/dataset.py [hunk 2]
```diff
@@ -74,20 +77,121 @@ def __getitem__(
...
+        if torch.is_tensor(input_ids):
+            input_list = input_ids.tolist()
+        else:
+            input_list = list(input_ids)
+        curr_len = len(input_list)
+
+        sep_id = self.sep_token_id
+        if sep_id in input_list:
+            sep_idx = input_list.index(sep_id)
+            new_labels = ([-100] * (sep_idx + 1)) + input_list[sep_idx + 1:]
```
Comment (Jatewo) [inline src/data/dataset.py]: Do we want to mask the ciphertext or let the model predict this first like the other models?

File: test/test_eval.py
Comment (Jatewo) [inline test/test_eval.py:1]: Lots of AI comments

File: src/engine/trainer.py [hunk 3]
```diff
@@ -173,13 +215,13 @@ def train(self, epochs: int) -> None:
...
         for epoch in range(start_epoch, epochs):
-            self.current_epoch = epoch + 1
+            self.current_epoch = epoch
 
             avg_train_loss = self._train_one_epoch()
             avg_val_loss = self._validate_one_epoch()
 
             current_lr = self.optimizer.param_groups[0]["lr"]
-            self.scheduler.step(avg_val_loss)
+            #self.scheduler.step(avg_val_loss)
```
Comment (Jatewo) [inline src/engine/trainer.py]: Commented-out code

File: src/data/dataset.py [hunk 4]
```diff
@@ -74,20 +77,121 @@ def __getitem__(
...
+        """
+        self.dataset = load_from_disk(str(directory_path))
+        self.max_seq_len = max_seq_len
+        self.sep_token_id = config.unique_homophones + 1
+
+    def __len__(self) -> int:
+        """Return the total number of samples in the dataset."""
+        return len(self.dataset)
+
+    def __getitem__(self, idx: int) -> dict[torch.Tensor, torch.Tensor]:
```
Comment (Jatewo) [inline src/data/dataset.py]: The typing is not correct here. The dict that is returned is:
```python
{
    "input_ids": tensor
    "labels": tensor
}
```

I.e. `dict[str, torch.Tensor]` since the keys are strings.

File: src/data/tokenizer.py [hunk 5]
```diff
@@ -6,30 +6,41 @@ class CipherTokenizer:
...
         self.config = config
         self.pad_token_id = 0
-        self.sep_token = config.unique_homophones + 1
-        self.char_offset = self.sep_token + 1
 
-        self.char_to_id = {chr(i + 97): i for i in range(26)}
-        self.id_to_char = {i + self.char_offset: chr(i + 97) for i in range(26)}
+        self.sep_token_id = config.unique_homophones + 1
+        self.space_token_id = config.unique_homophones + 2
+        self.eos_token_id = self.config.unique_homophones + 4
```
Comment (Jatewo) [inline src/data/tokenizer.py]: Why do we skip +3?

File: src/data/dataset.py [hunk 6]
```diff
@@ -74,20 +77,121 @@ def __getitem__(
...
+        curr_len = len(input_list)
+
+        sep_id = self.sep_token_id
+        if sep_id in input_list:
+            sep_idx = input_list.index(sep_id)
+            new_labels = ([-100] * (sep_idx + 1)) + input_list[sep_idx + 1:]
+        else:
+            new_labels = list(item["labels"])
+
+        pad_len = self.max_seq_len - curr_len
```
Comment (Jatewo) [inline src/data/dataset.py]: I don't think we should add padding here. This also defeats the purpose of the dynamic `safe_pad_collate`method in the data manager.

File: src/data/dataset.py
Comment (Jatewo) [inline src/data/dataset.py]: Why do we have two classes for this? All the data should be uniform arrow files.

File: src/engine/solver.py [hunk 7]
```diff
@@ -78,39 +79,66 @@ def load_checkpoint(self, checkpoint_path: Path) -> "CipherSolver":
         return self
 
     @torch.no_grad()
-    def decrypt(self, ciphertext: list[int] | str) -> str:
-        """High-level API: Take raw ciphertext and return a human-readable string.
+    def decrypt(self, input_ids: str | list[int] | torch.Tensor) -> str:
```
Comment (Jatewo) [inline src/engine/solver.py:82]: This function could really do with some refactoring, distributing some of the logic to helpers. It is very complex right now.

File: src/engine/trainer.py [hunk 8]
```diff
@@ -41,6 +42,7 @@ def __init__(
         val_loader: DataLoader,
         config: Config,
         save_path: Path,
+        run_type: str = "normal",
         exp_dir: Path = None,
         device: str = "cuda",
```
Comment (Jatewo) [inline src/engine/trainer.py:45]: Maybe add some of these to the config rather than also having a ton of separate arguments?

File: src/engine/trainer.py [hunk 9]
```diff
@@ -197,46 +239,151 @@ def train(self, epochs: int) -> None:
...
                 self.best_val_loss = avg_val_loss
 
+            self.current_epoch = epoch + 1
             self._save_checkpoint(avg_val_loss, is_best)
 
             if current_lr < 1e-7:
                 logger.info("Learning rate too low. Stopping early.")
                 break
 
     def _train_one_epoch(self) -> float:
```
Comment (Jatewo) [inline src/engine/trainer.py:240]: This is one of the methods I have ever seen. Definitely needs some helpers or distribution of the logic. It is both complex and deeply nested.

File: src/utils/data_manager.py [hunk 10]
```diff
@@ -2,162 +2,214 @@
...
+			for entry in tqdm(entries, desc="Indexing files", leave=False):
+				if entry.is_file():
+					if entry.name.endswith(".json"):
+						file_paths.append((entry.path, None))
+					elif entry.name.endswith(".zip"):
+						with zipfile.ZipFile(entry.path, "r") as z:
+							for name in z.namelist():
+								if name.endswith(".json"):
+									file_paths.append((entry.path, name))
+		return file_paths
```
Comment (Jatewo) [inline src/utils/data_manager.py]: Why are we looking for json files?

File: pyproject.toml [hunk 11]
```diff
@@ -62,4 +64,4 @@ ignore = [
 	"ERA001", # Found commented-out code
 ]
 [tool.ruff.format]
-indent-style = "tab"
\ No newline at end of file
+indent-style = "tab"
```
Comment (Jatewo) [inline pyproject.toml]: Didn't we change this to space?

**Review by Jatewo — CHANGES_REQUESTED** [review]

File: src/engine/solver.py (hunk 1 above)
Comment (MaizeHH) [reply src/engine/solver.py]: this is because input_ids at that point includes [BOS] [Cipher] [SEP], so it just takes the length of the cipher without the BOs and SEP tokens

File: src/data/tokenizer.py (hunk 5 above)
Comment (MaizeHH) [reply src/data/tokenizer.py]: i didn't find the need to explicitly use the BOS token, like i did with the other tokens

File: src/data/dataset.py (hunk 2 above)
Comment (MaizeHH) [reply src/data/dataset.py]: think we agreed previously not to mask the ciphertext, so ive removed it now 👍

File: src/engine/solver.py (hunk 1 above)
Comment (Jatewo) [reply src/engine/solver.py]: But do we ever define [BOS]?

File: src/data/dataset.py (hunk 6 above)
Comment (MaizeHH) [reply src/data/dataset.py]: i have removed the fixed padding across the entire dataset. it now relies purely on safe_pad_collate, which does dynamic padding (pads based on longest sequence per batch)

File: src/engine/solver.py (hunk 7 above)
Comment (MaizeHH) [reply src/engine/solver.py:82]: it has been changed now with the use of 2 helper functions

File: src/engine/trainer.py (hunk 9 above)
Comment (MaizeHH) [reply src/engine/trainer.py:240]: this has also been refactored with the use of 3 helper functions

File: src/data/dataset.py [hunk 12]
```diff
@@ -1,93 +1,81 @@
...
-                }
-                return cipher_tensor, data["plaintext"], metadata
+        if torch.is_tensor(input_ids):
+            input_list = input_ids.tolist()
+        else:
+            input_list = list(input_ids)
 
-            encoded_plain = self.tokenizer.encode(data["plaintext"])
-            plain_tensor = self.tokenizer.pad_sequence(encoded_plain, self.max_seq_len)
+        label_list = labels.tolist() if torch.is_tensor(labels) else list(labels)
```
Comment (Jatewo) [inline src/data/dataset.py]: isinstance() is better for type checking/annotation:
```suggestion
        label_list = (
            labels.tolist() if isinstance(labels, torch.Tensor) else list(labels)
        )
```

Note that this is literally the implementation of is_tensor():
```python
def is_tensor(obj: Any) -> TypeIs[Tensor]:
    r"""Returns True if `obj` is a PyTorch tensor.

    Note that this function is simply doing ``isinstance(obj, Tensor)``.
    Using that ``isinstance`` check is better for typechecking with mypy,
    and more explicit - so it's recommended to use that instead of ``is_tensor``.
    ...
    """
    return isinstance(obj, torch.Tensor)
```

File: src/data/dataset.py [hunk 13]
```diff
@@ -1,93 +1,81 @@
...
+        input_ids = item["input_ids"]
+        labels = item["labels"]
 
-            if self.mode == "eval":
-                metadata = {
-                    "path": str(path),
-                    "internal_name": internal_name if internal_name is not None else "",
-                }
-                return cipher_tensor, data["plaintext"], metadata
+        if torch.is_tensor(input_ids):
```
Comment (Jatewo) [inline src/data/dataset.py]: isinstance() is better for type checking/annotation:
```suggestion
        input_list = (
            input_ids.tolist()
            if isinstance(input_ids, torch.Tensor)
            else list(input_ids)
        )
```
Note that this is literally the implementation of is_tensor():
```python
def is_tensor(obj: Any) -> TypeIs[Tensor]:
    r"""Returns True if `obj` is a PyTorch tensor.

    Note that this function is simply doing ``isinstance(obj, Tensor)``.
    Using that ``isinstance`` check is better for typechecking with mypy,
    and more explicit - so it's recommended to use that instead of ``is_tensor``.
    ...
    """
    return isinstance(obj, torch.Tensor)
```

**Review by Jatewo — CHANGES_REQUESTED** [review]

File: src/engine/trainer.py (hunk 8 above)
Comment (MaizeHH) [reply src/engine/trainer.py:45]: i have removed save_path from the arguments as it was just passed save_path from the config anyway

File: src/engine/trainer.py [hunk 14]
```diff
@@ -61,32 +62,59 @@ def __init__(
...
 
         if exp_dir:
             self.exp_dir = exp_dir
             self.timestamp = exp_dir.name.replace("exp_", "")
         else:
             self.timestamp = datetime.now().strftime("%d%m_%H%M_%Y")
-            self.exp_dir = save_path / f"exp_{self.timestamp}"
+            self.exp_dir = config.save_path / f"exp_{self.run_type}_{self.timestamp}"
             self.exp_dir.mkdir(parents=True, exist_ok=True)
             self._save_config()
```
Comment (Jatewo) [inline src/engine/trainer.py]: Remove exp_dir as a parameter and let it handle it internally:
```suggestion
        self.timestamp = datetime.now().strftime("%d%m_%H%M_%Y")
        self.exp_dir = self.save_path / f"exp_{self.run_type}_{self.timestamp}"
        self.exp_dir.mkdir(parents=True, exist_ok=True)
        self._save_config()
```

File: src/engine/trainer.py [hunk 15]
```diff
@@ -40,7 +41,7 @@ def __init__(
         train_loader: DataLoader,
         val_loader: DataLoader,
         config: Config,
-        save_path: Path,
+        run_type: str = "normal",
         exp_dir: Path = None,
```
Comment (Jatewo) [inline src/engine/trainer.py]: Remove this
```suggestion
```

**Review by Jatewo — CHANGES_REQUESTED** [review]

File: src/engine/trainer.py [hunk 16]
```diff
@@ -101,17 +129,23 @@ def _save_history(self) -> None:
             json.dump(self.history, f, indent=4)
         logger.info(f"History updated at {history_path}")
 
-    def _save_checkpoint(self, val_loss: float, is_best: bool) -> None:
-        """Save a model checkpoint and update the 'latest' and 'best' files.
+    def _save_checkpoint(
+        self,
+        val_loss: float,
+        is_best: bool,
+        suffix: str = None,
```
Comment (Jatewo) [inline src/engine/trainer.py]:
```suggestion
        suffix: str | None = None,
```

**Review by Jatewo — CHANGES_REQUESTED** [review]

File: src/engine/solver.py [hunk 17]
```diff
@@ -78,39 +79,108 @@ def load_checkpoint(self, checkpoint_path: Path) -> "CipherSolver":
...
+        logits = self.model(input_ids, inference_params=inference_params)
+        next_token = torch.argmax(logits[:, -1, :], dim=-1).view(1, 1)
+
+        generated_tokens = []
+        target_len = input_ids.size(1) - 2
+
+        for _ in range(target_len):
+            token_id = next_token.item()
+            if token_id == self.tokenizer.eos_token_id:
+                break
```
Comment (Jatewo) [inline src/engine/solver.py:187]: Right now an error will be raised if `self.model` is `None`. 

Perhaps we want to either fail more gracefully (logging an error/warning to the user and returning without raising) or at least logging the error and then raise an error from there.

Comment (Jatewo) [reply src/engine/solver.py:187]: Since we do already have a check in the main method, it may not make sense to do another check here. Rather you could make the helper a static method and pass the model (and token_id) instead.
```suggestion
    @staticmethod
    def _generate_autoregressive(
        model: MambaModel,
        eos_token_id: int,
        input_ids: torch.Tensor,
        inference_params: InferenceParams,
    ) -> list[int]:
        """Execute the Mamba autoregressive loop using provided inference parameters.

        Args:
            model: The Mamba neural network model.
            eos_token_id: The token ID used to signal the end of a sequence.
            input_ids: The pre-processed input tensor [1, seq_len] containing
                the ciphertext and the separator token.
            inference_params: A Mamba `InferenceParams` object that tracks
                the internal SSM states across time steps for efficient generation.

        Returns:
            list[int]: A list of generated token IDs representing the decrypted
                plaintext.

        """
        logits = model(input_ids, inference_params=inference_params)
        next_token = torch.argmax(logits[:, -1, :], dim=-1).view(1, 1)

        generated_tokens = []
        target_len = input_ids.size(1) - 2

        for _ in range(target_len):
            token_id = next_token.item()
            if token_id == eos_token_id:
                break

            generated_tokens.append(token_id)
            logits = model(next_token, inference_params=inference_params)
            next_token = torch.argmax(logits[:, -1, :], dim=-1).view(1, 1)

        return generated_tokens
```

**Review by Jatewo — CHANGES_REQUESTED** [review]

File: src/engine/solver.py [hunk 18]
```diff
@@ -78,39 +79,108 @@ def load_checkpoint(self, checkpoint_path: Path) -> "CipherSolver":
...
-            .unsqueeze(0)
-            .to(self.device)
+        inference_params = InferenceParams(
+            max_seqlen=self.config.max_len,
+            max_batch_size=1,
         )
 
-        logits = self.model(input_tensor)
-        pred_indices = torch.argmax(logits, dim=-1).squeeze(0).tolist()
+        generated_tokens = self._generate_autoregressive(input_ids, inference_params)
```
Comment (Jatewo) [inline src/engine/solver.py]: Assuming the changes to the helper were made:
```suggestion
        generated_tokens = self._generate_autoregressive(
            model=self.model,
            eos_token_id=self.tokenizer.eos_token_id,
            inference_params=inference_params,
            input_ids=input_ids,
        )
```

**Review by Jatewo — CHANGES_REQUESTED** [review]

**Review by Jatewo — APPROVED** [review]
LGTM

**Review by TheMOHD3886 — APPROVED** [review]
lgtm 🌊
